File: backend/internal/service/fingerprint_test.go
package service

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestNewDeviceFingerprintService(t *testing.T) {
	service := NewDeviceFingerprintService()
	assert.NotNil(t, service)
}

func TestGenerateFingerprintHash(t *testing.T) {
	service := NewDeviceFingerprintService()

	data := FingerprintData{
		UserAgent:           "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36",
		ScreenWidth:         1920,
		ScreenHeight:        1080,
		ColorDepth:          24,
		Timezone:           "Asia/Shanghai",
		Language:           "zh-CN",
		Platform:           "Win32",
		HardwareConcurrency: 8,
		DeviceMemory:       8,
		TouchPoints:        0,
		WebGLVendor:        "Google Inc.",
		WebGLRenderer:      "ANGLE (Intel, Intel(R) UHD Graphics Direct3D11)",
		CanvasFingerprint:  "test-canvas-hash",
		AudioFingerprint:   "test-audio-hash",
		Fonts:              []string{"Arial", "Helvetica"},
		Plugins:            []string{"Plugin1", "Plugin2"},
	}

	hash := service.GenerateFingerprintHash(data)

	assert.NotEmpty(t, hash.UserAgentHash)
	assert.NotEmpty(t, hash.ScreenHash)
	assert.NotEmpty(t, hash.BrowserHash)
	assert.NotEmpty(t, hash.PlatformHash)
	assert.NotEmpty(t, hash.CanvasHash)
	assert.NotEmpty(t, hash.WebGLHash)
	assert.NotEmpty(t, hash.AudioHash)
}

func TestGenerateFingerprintHashEmptyCanvas(t *testing.T) {
	service := NewDeviceFingerprintService()

	data := FingerprintData{
		UserAgent:   "Mozilla/5.0",
		Platform:    "Win32",
		WebGLRenderer: "",
	}

	hash := service.GenerateFingerprintHash(data)

	assert.NotEmpty(t, hash.CanvasHash)
}

func TestGenerateFingerprintHashEmptyAudio(t *testing.T) {
	service := NewDeviceFingerprintService()

	data := FingerprintData{
		UserAgent:   "Mozilla/5.0",
		Platform:    "Win32",
		Language:   "en",
		DeviceMemory: 8,
	}

	hash := service.GenerateFingerprintHash(data)

	assert.NotEmpty(t, hash.AudioHash)
}

func TestCombineHashes(t *testing.T) {
	service := NewDeviceFingerprintService()

	hash := FingerprintHash{
		UserAgentHash: "abc123",
		ScreenHash:    "def456",
		BrowserHash:   "ghi789",
		PlatformHash:  "jkl012",
		CanvasHash:    "mno345",
		WebGLHash:     "pqr678",
		AudioHash:     "stu901",
	}

	combined := service.CombineHashes(hash)

	assert.NotEmpty(t, combined)
	assert.Len(t, combined, 64)
}

func TestHashString(t *testing.T) {
	service := NewDeviceFingerprintService()

	hash1 := service.hashString("test-input")
	hash2 := service.hashString("test-input")
	hash3 := service.hashString("different-input")

	assert.Equal(t, hash1, hash2)
	assert.NotEqual(t, hash1, hash3)
	assert.Len(t, hash1, 64)
}

func TestCalculateSimilarity(t *testing.T) {
	service := NewDeviceFingerprintService()

	tests := []struct {
		name     string
		hash1    string
		hash2    string
		expected float64
	}{
		{
			name:     "identical hashes",
			hash1:    "abcdef1234567890",
			hash2:    "abcdef1234567890",
			expected: 1.0,
		},
		{
			name:     "completely different hashes",
			hash1:    "0000000000000000",
			hash2:    "ffffffffffffffff",
			expected: 0.0,
		},
		{
			name:     "partial match",
			hash1:    "abcdef1234567890",
			hash2:    "abcdef0000000000",
			expected: 0.5,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := service.CalculateSimilarity(tt.hash1, tt.hash2)
			assert.Equal(t, tt.expected, result)
		})
	}
}

func TestCalculateSimilarityDifferentLengths(t *testing.T) {
	service := NewDeviceFingerprintService()

	result := service.CalculateSimilarity("short", "muchlongerhash")
	assert.Equal(t, 0.0, result)
}

func TestFingerprintWeights(t *testing.T) {
	expectedWeights := map[string]float64{
		"user_agent": 1.5,
		"screen":     1.2,
		"browser":    1.3,
		"platform":   1.0,
		"canvas":     2.0,
		"webgl":      1.8,
		"audio":      1.5,
	}

	for key, expected := range expectedWeights {
		actual, exists := fingerprintWeights[key]
		assert.True(t, exists, "Weight for %s should exist", key)
		assert.Equal(t, expected, actual)
	}
}

func TestFingerprintDataStructure(t *testing.T) {
	data := FingerprintData{
		UserAgent:           "Mozilla/5.0",
		ScreenWidth:         1920,
		ScreenHeight:        1080,
		ColorDepth:          24,
		Timezone:           "UTC+8",
		Language:           "en-US",
		Platform:           "MacIntel",
		HardwareConcurrency: 4,
		DeviceMemory:       8,
		TouchPoints:        0,
		WebGLVendor:        "Apple Inc.",
		WebGLRenderer:      "Apple GPU",
		CanvasFingerprint:  "canvas123",
		AudioFingerprint:   "audio456",
		Fonts:              []string{"Arial"},
		Plugins:            []string{"PDF Viewer"},
		DoNotTrack:         true,
		CookiesEnabled:     true,
		LocalStorage:       true,
		SessionStorage:     true,
	}

	assert.Equal(t, "Mozilla/5.0", data.UserAgent)
	assert.Equal(t, 1920, data.ScreenWidth)
	assert.Equal(t, 1080, data.ScreenHeight)
	assert.True(t, data.DoNotTrack)
	assert.True(t, data.CookiesEnabled)
}

func TestFingerprintHashStructure(t *testing.T) {
	hash := FingerprintHash{
		UserAgentHash: "ua-hash",
		ScreenHash:    "screen-hash",
		BrowserHash:   "browser-hash",
		PlatformHash:  "platform-hash",
		CanvasHash:    "canvas-hash",
		WebGLHash:     "webgl-hash",
		AudioHash:     "audio-hash",
	}

	assert.NotEmpty(t, hash.UserAgentHash)
	assert.NotEmpty(t, hash.ScreenHash)
	assert.NotEmpty(t, hash.BrowserHash)
	assert.NotEmpty(t, hash.PlatformHash)
	assert.NotEmpty(t, hash.CanvasHash)
	assert.NotEmpty(t, hash.WebGLHash)
	assert.NotEmpty(t, hash.AudioHash)
}

func TestRiskAssessmentStructure(t *testing.T) {
	assessment := RiskAssessment{
		Score:          45.5,
		Level:          "medium",
		Factors:        []string{"factor1", "factor2"},
		IsNewDevice:    true,
		IsSharedDevice: false,
		Similarity:     0.85,
	}

	assert.Equal(t, 45.5, assessment.Score)
	assert.Equal(t, "medium", assessment.Level)
	assert.Len(t, assessment.Factors, 2)
	assert.True(t, assessment.IsNewDevice)
	assert.False(t, assessment.IsSharedDevice)
	assert.Equal(t, 0.85, assessment.Similarity)
}

func TestCollectedFingerprintStructure(t *testing.T) {
	fingerprint := CollectedFingerprint{
		FingerprintID: 123,
		Hash:         "hash-value",
		RiskLevel:    "low",
	}

	assert.Equal(t, uint(123), fingerprint.FingerprintID)
	assert.Equal(t, "hash-value", fingerprint.Hash)
	assert.Equal(t, "low", fingerprint.RiskLevel)
}

func TestDeviceInfoStructure(t *testing.T) {
	info := DeviceInfo{
		ID:           456,
		Hash:         "device-hash",
		UserAgent:    "Mozilla/5.0",
		ScreenInfo:   "1920x1080",
		BrowserInfo:  "Chrome",
		PlatformInfo: "Windows",
		VisitCount:   10,
		IsTrusted:    true,
		RiskLevel:    "low",
	}

	assert.Equal(t, uint(456), info.ID)
	assert.Equal(t, "device-hash", info.Hash)
	assert.Equal(t, 10, info.VisitCount)
	assert.True(t, info.IsTrusted)
}

func TestSimilarDeviceStructure(t *testing.T) {
	similar := SimilarDevice{
		DeviceID:   789,
		Similarity: 0.92,
		VisitCount: 5,
	}

	assert.Equal(t, uint(789), similar.DeviceID)
	assert.Equal(t, 0.92, similar.Similarity)
	assert.Equal(t, 5, similar.VisitCount)
}

func TestRiskLevels(t *testing.T) {
	levels := []string{"low", "medium", "high"}

	for _, level := range levels {
		assert.Contains(t, []string{"low", "medium", "high"}, level)
	}
}

func TestFingerprintHashConsistency(t *testing.T) {
	service := NewDeviceFingerprintService()

	data := FingerprintData{
		UserAgent:     "TestAgent",
		Platform:      "TestPlatform",
		WebGLRenderer: "TestRenderer",
	}

	hash1 := service.GenerateFingerprintHash(data)
	hash2 := service.GenerateFingerprintHash(data)

	assert.Equal(t, hash1.UserAgentHash, hash2.UserAgentHash)
	assert.Equal(t, hash1.ScreenHash, hash2.ScreenHash)
	assert.Equal(t, hash1.BrowserHash, hash2.BrowserHash)
	assert.Equal(t, hash1.PlatformHash, hash2.PlatformHash)
	assert.Equal(t, hash1.CanvasHash, hash2.CanvasHash)
	assert.Equal(t, hash1.WebGLHash, hash2.WebGLHash)
	assert.Equal(t, hash1.AudioHash, hash2.AudioHash)
}

func TestFingerprintHashUniqueness(t *testing.T) {
	service := NewDeviceFingerprintService()

	data1 := FingerprintData{
		UserAgent:     "Mozilla/5.0 Chrome",
		Platform:      "Win32",
		WebGLRenderer: "Renderer1",
		ScreenWidth:   1920,
		ScreenHeight:  1080,
	}

	data2 := FingerprintData{
		UserAgent:     "Mozilla/5.0 Firefox",
		Platform:      "Win32",
		WebGLRenderer: "Renderer2",
		ScreenWidth:   1920,
		ScreenHeight:  1080,
	}

	hash1 := service.GenerateFingerprintHash(data1)
	hash2 := service.GenerateFingerprintHash(data2)

	assert.NotEqual(t, hash1.UserAgentHash, hash2.UserAgentHash)
	assert.NotEqual(t, hash1.BrowserHash, hash2.BrowserHash)
	assert.NotEqual(t, hash1.WebGLHash, hash2.WebGLHash)
}

File: frontend/static/js/tests/fingerprint.test.js
describe('FingerprintCollector', () => {
    let collector;

    beforeEach(() => {
        collector = new FingerprintCollector();
    });

    describe('Constructor', () => {
        test('should initialize with empty cache and components', () => {
            expect(collector.cache).toBeNull();
            expect(collector.components).toEqual({});
        });
    });

    describe('Hash String Generation', () => {
        test('should generate consistent hash for same input', async () => {
            const input = 'test-string-123';
            
            const hash1 = await collector.hashString(input);
            const hash2 = await collector.hashString(input);
            
            expect(hash1).toBe(hash2);
        });

        test('should generate different hashes for different inputs', async () => {
            const hash1 = await collector.hashString('input1');
            const hash2 = await collector.hashString('input2');
            
            expect(hash1).not.toBe(hash2);
        });

        test('should generate SHA-256 hash format', async () => {
            const hash = await collector.hashString('test');
            
            expect(hash).toMatch(/^[a-f0-9]{64}$/);
        });
    });

    describe('Combined Hash Generation', () => {
        test('should generate combined hash from fingerprint data', async () => {
            const fingerprintData = {
                user_agent: 'Mozilla/5.0 Test Browser',
                screen_width: 1920,
                screen_height: 1080,
                canvas_fingerprint: 'canvas-hash-123',
                webgl_vendor: 'Vendor',
                webgl_renderer: 'Renderer',
                audio_fingerprint: 'audio-hash-456',
                platform: 'Win32'
            };

            const combinedHash = await collector.generateCombinedHash(fingerprintData);

            expect(combinedHash).toMatch(/^[a-f0-9]{64}$/);
        });

        test('should generate different combined hashes for different data', async () => {
            const data1 = {
                user_agent: 'Browser1',
                screen_width: 1920,
                screen_height: 1080,
                canvas_fingerprint: 'hash1',
                webgl_vendor: 'V1',
                webgl_renderer: 'R1',
                audio_fingerprint: 'ah1',
                platform: 'Win32'
            };

            const data2 = {
                user_agent: 'Browser2',
                screen_width: 1920,
                screen_height: 1080,
                canvas_fingerprint: 'hash2',
                webgl_vendor: 'V2',
                webgl_renderer: 'R2',
                audio_fingerprint: 'ah2',
                platform: 'Win32'
            };

            const hash1 = await collector.generateCombinedHash(data1);
            const hash2 = await collector.generateCombinedHash(data2);

            expect(hash1).not.toBe(hash2);
        });
    });

    describe('Cache Management', () => {
        test('should return cached data on subsequent calls', async () => {
            const data = await collector.collect();
            const cachedData = await collector.collect();

            expect(cachedData).toEqual(data);
        });

        test('should clear cache correctly', async () => {
            await collector.collect();
            expect(collector.cache).not.toBeNull();

            collector.clearCache();
            expect(collector.cache).toBeNull();
        });

        test('should clear components on cache clear', async () => {
            await collector.collect();
            expect(Object.keys(collector.components).length).toBeGreaterThan(0);

            collector.clearCache();
            expect(collector.components).toEqual({});
        });
    });

    describe('Component Retrieval', () => {
        test('should return collected components', async () => {
            await collector.collect();
            
            const components = collector.getComponents();
            
            expect(components).toHaveProperty('userAgent');
            expect(components).toHaveProperty('screen');
            expect(components).toHaveProperty('browser');
        });
    });

    describe('Data Collection Structure', () => {
        test('should return complete fingerprint data structure', async () => {
            const data = await collector.collect();

            expect(data).toHaveProperty('user_agent');
            expect(data).toHaveProperty('screen_width');
            expect(data).toHaveProperty('screen_height');
            expect(data).toHaveProperty('color_depth');
            expect(data).toHaveProperty('timezone');
            expect(data).toHaveProperty('language');
            expect(data).toHaveProperty('platform');
            expect(data).toHaveProperty('hardware_concurrency');
            expect(data).toHaveProperty('device_memory');
            expect(data).toHaveProperty('touch_points');
            expect(data).toHaveProperty('webgl_vendor');
            expect(data).toHaveProperty('webgl_renderer');
            expect(data).toHaveProperty('canvas_fingerprint');
            expect(data).toHaveProperty('audio_fingerprint');
            expect(data).toHaveProperty('fonts');
            expect(data).toHaveProperty('plugins');
            expect(data).toHaveProperty('do_not_track');
            expect(data).toHaveProperty('cookies_enabled');
            expect(data).toHaveProperty('local_storage');
            expect(data).toHaveProperty('session_storage');
            expect(data).toHaveProperty('collection_time_ms');
        });

        test('should include collection time', async () => {
            const data = await collector.collect();

            expect(data.collection_time_ms).toBeGreaterThanOrEqual(0);
        });
    });
});

describe('FingerprintAPI', () => {
    let api;

    beforeEach(() => {
        api = new FingerprintAPI();
    });

    describe('Constructor', () => {
        test('should initialize with FingerprintCollector', () => {
            expect(api.collector).toBeInstanceOf(FingerprintCollector);
        });

        test('should set default API base URL', () => {
            expect(api.apiBase).toBe('/api/v1/fingerprint');
        });
    });

    describe('Token Retrieval', () => {
        test('should retrieve token from localStorage', () => {
            localStorage.setItem('token', 'test-token-123');

            const token = api.getToken();

            expect(token).toBe('test-token-123');

            localStorage.removeItem('token');
        });

        test('should retrieve token from sessionStorage as fallback', () => {
            sessionStorage.setItem('token', 'session-token');

            const token = api.getToken();

            expect(token).toBe('session-token');

            sessionStorage.removeItem('token');
        });

        test('should return empty string when no token found', () => {
            localStorage.removeItem('token');
            sessionStorage.removeItem('token');

            const token = api.getToken();

            expect(token).toBe('');
        });
    });

    describe('API Response Structure', () => {
        test('should handle successful response structure', () => {
            const mockResult = {
                success: true,
                fingerprintId: 123,
                hash: 'abc123',
                riskLevel: 'low',
                collectionTime: 50
            };

            expect(mockResult).toHaveProperty('success');
            expect(mockResult).toHaveProperty('fingerprintId');
            expect(mockResult).toHaveProperty('hash');
            expect(mockResult).toHaveProperty('riskLevel');
        });

        test('should handle error response structure', () => {
            const mockError = {
                success: false,
                error: 'Network error'
            };

            expect(mockError.success).toBe(false);
            expect(mockError).toHaveProperty('error');
        });
    });
});

describe('Device Management', () => {
    describe('Device Info Structure', () => {
        test('should validate device info structure', () => {
            const deviceInfo = {
                id: 1,
                hash: 'device-hash-123',
                user_agent: 'Mozilla/5.0',
                screen_info: '1920x1080',
                browser_info: 'Chrome',
                platform_info: 'Windows',
                visit_count: 10,
                is_trusted: true,
                risk_level: 'low'
            };

            expect(deviceInfo).toHaveProperty('id');
            expect(deviceInfo).toHaveProperty('hash');
            expect(deviceInfo).toHaveProperty('visit_count');
            expect(deviceInfo).toHaveProperty('is_trusted');
            expect(deviceInfo).toHaveProperty('risk_level');
        });

        test('should validate risk levels', () => {
            const validLevels = ['low', 'medium', 'high'];

            expect(validLevels).toContain('low');
            expect(validLevels).toContain('medium');
            expect(validLevels).toContain('high');
        });
    });

    describe('Similar Device Structure', () => {
        test('should validate similar device structure', () => {
            const similarDevice = {
                device_id: 1,
                similarity: 0.92,
                first_seen_at: '2024-01-01T00:00:00Z',
                last_seen_at: '2024-01-15T00:00:00Z',
                visit_count: 5
            };

            expect(similarDevice).toHaveProperty('device_id');
            expect(similarDevice).toHaveProperty('similarity');
            expect(similarDevice.similarity).toBeGreaterThanOrEqual(0);
            expect(similarDevice.similarity).toBeLessThanOrEqual(1);
        });
    });
});
